### ISSUE TITLE
Similar results are obtained simply by using `torch.compile`

### ISSUE THREAD

**Issue by rohitrango** [xi-jia/FastLNCC#1]
Hi,

I havent had time to verify the memory footprint of the method (I assume it will be the same, since there is no redundant computation that is reduced). 

However, you can get the same results in terms of time by simply wrapping the computation with a `torch.compile` 

It would be nice to add that comparison

**Comment by xi-jia** [comment]
To our knowledge, torch.compile has some compatibility problems with older versions of torch (<2.0) and certain GPUs such as the P100 used in our small report. But, yes, torch.compile can be a feasible solution if both the torch version and GPU cards support it.

In our experiments, we found that on the A100, when `torch.backends.cudnn.allow_tf32 = True`, the training time of torch.compile (using the default Triton mode, excluding compile time) is **faster** than ours. However, allowing `tf32` can make training with torch.compile difficult to converge: multiple models initialized with different seeds **fail to converge**, whereas ours do. When `torch.backends.cudnn.allow_tf32 = False`, torch.compile retains convergence but presents a **similar** training time with ours--both being slower than them of `allow_tf32 = True`.

On the RTX8000, we found that for smaller models like VM, torch.compile is **slower** than our method, regardless of whether allow_tf32 is enabled. For larger models like TM, torch.compile _can be faster_ if more epochs are involved.

Overall, neither is perfect; one can decide which to use based on the available resources.
Thank you for the discussion—I’ll label this thread for reference.

**Comment by rohitrango** [comment]
Very interesting discussion! I think this code can be wrapped in a scheduler that benchmarks the different LNCC variants upon script startup and chooses the best one based on runtime (since all of them produce identical results). This will significantly improve aggregate runtime for many training scripts.

The convergence problem with allowing `tf32` is new to me, I will look into this.
